File: app/src/main/java/io/treehouses/remote/adapter/ServicesListAdapter.java
package io.treehouses.remote.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.Collections;

import io.treehouses.remote.R;
import io.treehouses.remote.pojo.ServiceInfo;

public class ServicesListAdapter extends ArrayAdapter<ServiceInfo> {
    private ArrayList<ServiceInfo> data;
    private Context context;
    private TextView name;
    private ImageView status;
    private int headerColour;
    //private Button start, install, restart, link, info;

    public ServicesListAdapter(Context context, ArrayList<ServiceInfo> services, int headerColour) {
        super(context,0,services);
        this.data = services;
        this.context = context;
        this.headerColour = headerColour;
    }

    @Override
    public ServiceInfo getItem(int position) {
        return data.get(position);
    }

    @Override
    public int getPosition(ServiceInfo s) {
        for (int i = 0; i < data.size(); i++) {
            if (data.get(i).name.equals(s.name) && data.get(i).serviceStatus == s.serviceStatus) {
                return i;
            }
        }
        return -1;
    }

    @Override
    public void notifyDataSetChanged() {
        Collections.sort(data);
        super.notifyDataSetChanged();
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        return initView(position, convertView, parent);
    }

    @Override
    public View getDropDownView(int position, View convertView, ViewGroup parent) {
        return initView(position, convertView, parent);
    }

    @Override
    public boolean isEnabled(int position){
        return data.get(position).serviceStatus != ServiceInfo.SERVICE_HEADER_INSTALLED && data.get(position).serviceStatus != ServiceInfo.SERVICE_HEADER_AVAILABLE;
    }


    private void findViews(View view) {
        name = view.findViewById(R.id.service_name);
        status = view.findViewById(R.id.service_status);
        name.setTextColor(headerColour);
//        start = view.findViewById(R.id.start_service);
//        install = view.findViewById(R.id.install_service);
//        restart = view.findViewById(R.id.restart_service);
//        link = view.findViewById(R.id.link_button);
//        info = view.findViewById(R.id.service_info);
    }

    private void setStatus(int statusCode) {
        if (statusCode == ServiceInfo.SERVICE_AVAILABLE) {

            //setButtons(false, false, false);

            name.setTextColor(context.getResources().getColor(R.color.md_grey_600));
            status.setImageDrawable(context.getResources().getDrawable(R.drawable.circle_red));
        } else if (statusCode == ServiceInfo.SERVICE_INSTALLED) {

            //setButtons(false, true, false);

            name.setTextColor(context.getResources().getColor(R.color.md_grey_600));
            status.setImageDrawable(context.getResources().getDrawable(R.drawable.circle_yellow));

        } else if (statusCode == ServiceInfo.SERVICE_RUNNING) {
            //setButtons(true, true, true);

            name.setTextColor(context.getResources().getColor(R.color.md_green_500));
            status.setImageDrawable(context.getResources().getDrawable(R.drawable.circle_green));

        }
    }

//    private void setOnClick(ViewGroup parent, View convertView, int id, int position) {
//        convertView.findViewById(id).setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View v) {
//                ((ListView) parent).performItemClick(v, position, 0);
//            }
//        });
//    }

    private View initView(int position, View convertView, ViewGroup parent) {
        if (data.get(position).serviceStatus != ServiceInfo.SERVICE_HEADER_AVAILABLE && data.get(position).serviceStatus != ServiceInfo.SERVICE_HEADER_INSTALLED) {
            convertView = LayoutInflater.from(context).inflate(R.layout.services_row_layout, parent, false);
        }
        else {
            convertView = LayoutInflater.from(context).inflate(R.layout.services_section_header, parent, false);

        }
        findViews(convertView);



        name.setText(data.get(position).name);

        setStatus(data.get(position).serviceStatus);

        return convertView;
    }
}

File: app/src/main/java/io/treehouses/remote/adapter/ServicesListAdapter.kt
package io.treehouses.remote.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.ImageView
import android.widget.TextView
import io.treehouses.remote.R
import io.treehouses.remote.pojo.ServiceInfo
import java.util.*

class ServicesListAdapter //private Button start, install, restart, link, info;
(private val mContext: Context, private val data: ArrayList<ServiceInfo>, private val headerColour: Int) : ArrayAdapter<ServiceInfo>(mContext, 0, data) {
    private var name: TextView? = null
    private var status: ImageView? = null
    override fun getItem(position: Int): ServiceInfo {
        return data[position]
    }

    override fun getPosition(item: ServiceInfo?): Int {
        for (i in data.indices) {
            if (data[i].name == item!!.name && data[i].serviceStatus == item!!.serviceStatus) {
                return i
            }
        }
        return -1
    }

    override fun notifyDataSetChanged() {
        Collections.sort(data)
        super.notifyDataSetChanged()
    }


    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        return initView(position, convertView, parent)
    }

    override fun getDropDownView(position: Int, convertView: View?, parent: ViewGroup): View {
        return initView(position, convertView, parent)
    }

    override fun isEnabled(position: Int): Boolean {
        return flag(position)
    }

    private fun findViews(view: View) {
        name = view.findViewById(R.id.service_name)
        status = view.findViewById(R.id.service_status)
        name!!.setTextColor(headerColour)
    }

    private fun setStatus(statusCode: Int) {
        var color:Int = 0
        var drawable:Int = 0
        if(statusCode == ServiceInfo.SERVICE_AVAILABLE || statusCode == ServiceInfo.SERVICE_INSTALLED || statusCode == ServiceInfo.SERVICE_RUNNING){
            if (statusCode == ServiceInfo.SERVICE_AVAILABLE) {
                color = R.color.md_grey_600
                drawable = R.drawable.circle_red
            } else if (statusCode == ServiceInfo.SERVICE_INSTALLED) {
                color = R.color.md_grey_600
                drawable = R.drawable.circle_yellow
            } else if (statusCode == ServiceInfo.SERVICE_RUNNING) {
                color = R.color.md_green_500
                drawable = R.drawable.circle_green
            }
            name!!.setTextColor(context.resources.getColor(color))
            status!!.setImageDrawable(context.resources.getDrawable(drawable))
        }

    }

    private fun initView(position: Int, convertView: View?, parent: ViewGroup): View {
        var convertView = convertView
        convertView = if (flag(position)) {
            LayoutInflater.from(context).inflate(R.layout.services_row_layout, parent, false)
        } else {
            LayoutInflater.from(context).inflate(R.layout.services_section_header, parent, false)
        }
        findViews(convertView)
        name!!.text = data[position].name
        setStatus(data[position].serviceStatus)
        return convertView
    }

    private fun flag(position:Int):Boolean {
        return data[position].serviceStatus != ServiceInfo.SERVICE_HEADER_AVAILABLE && data[position].serviceStatus != ServiceInfo.SERVICE_HEADER_INSTALLED
    }

}